[Cocoa] Text inserted after resized adaptive image glyph does not match font size

pxlcoder opened the pull request:
#### 0b81c7acf2f706c5ca0a33be8f5ddcb7ae511114
<pre>
[Cocoa] Text inserted after resized adaptive image glyph does not match font size
<a href="https://bugs.webkit.org/show_bug.cgi?id=278693">https://bugs.webkit.org/show_bug.cgi?id=278693</a>
<a href="https://rdar.apple.com/127761275">rdar://127761275</a>

Reviewed by Wenson Hsieh and Richard Robinson.

Adaptive image glyphs are essentially `&lt;picture&gt;` element with text-like behavior.
Traditional text insertion matches the font size of surrounding content as
created text nodes are inserted into a styled element.

However, when resizing an adaptive image glyph, the `font-size` property is
applied to the `&lt;picture&gt;` element. Then when inserting text after the
`&lt;picture&gt;` element, the text node is simply inserted as a sibling to the
`&lt;picture&gt;`, as `&lt;picture&gt;` cannot contain non-`&lt;source&gt;` or non-`&lt;img&gt;`
children due to editing. Consequently, the text takes on the size of the
common parent, rather than the adaptive image glyph.

Fix by getting an `EditingStyle` for the adaptive image glyph and applying
it to the inserted text.

* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::InsertTextCommand::doApply):

The approach taken here can easily be generalized to other inline elements.
However, to reduce risk and surface of the behavior change, restrict the fix to
adaptive image glyphs. Other browsers are also inconsistent in their behavior
here.

Additionally, remove a comment that simply describes what the code does.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertTextAfterAdaptiveImageGlyph)):

Canonical link: <a href="https://commits.webkit.org/282778@main">https://commits.webkit.org/282778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13c3b5eeb992df6d8736cf3462bbaba9d2f6d685

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43627 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16867 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68292 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14878 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66390 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51325 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15158 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51720 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10257 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67339 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40333 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55603 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32339 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12984 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13752 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58966 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13314 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69991 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12833 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59042 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55697 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59204 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6788 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/472 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9735 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39447 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40526 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41709 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40269 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/editing/InsertTextCommand.cpp:
```diff
@@ -230,8 +231,17 @@ void InsertTextCommand::doApply()
 
-    // Handle the case where there is a typing style.
-    if (RefPtr<EditingStyle> typingStyle = document().selection().typingStyle()) {
+    CheckedRef frameSelection = document().selection();
```
Is this CheckedRef safe? (since the code below seems to update layout)

pxlcoder replied:
Good catch, the code below does indeed update layout via `Element::computedStyle`.

Fixed.

pxlcoder committed:
[Cocoa] Text inserted after resized adaptive image glyph does not match font size
https://bugs.webkit.org/show_bug.cgi?id=278693
rdar://127761275

Reviewed by Wenson Hsieh and Richard Robinson.

Adaptive image glyphs are essentially `<picture>` element with text-like behavior.
Traditional text insertion matches the font size of surrounding content as
created text nodes are inserted into a styled element.

However, when resizing an adaptive image glyph, the `font-size` property is
applied to the `<picture>` element. Then when inserting text after the
`<picture>` element, the text node is simply inserted as a sibling to the
`<picture>`, as `<picture>` cannot contain non-`<source>` or non-`<img>`
children due to editing. Consequently, the text takes on the size of the
common parent, rather than the adaptive image glyph.

Fix by getting an `EditingStyle` for the adaptive image glyph and applying
it to the inserted text.

* Source/WebCore/editing/InsertTextCommand.cpp:
(WebCore::InsertTextCommand::doApply):

The approach taken here can easily be generalized to other inline elements.
However, to reduce risk and surface of the behavior change, restrict the fix to
adaptive image glyphs. Other browsers are also inconsistent in their behavior
here.

Additionally, remove a comment that simply describes what the code does.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/AdaptiveImageGlyph.mm:
(TestWebKitAPI::TEST(AdaptiveImageGlyph, InsertTextAfterAdaptiveImageGlyph)):

Canonical link: https://commits.webkit.org/282778@main
